Remote Schema Permissions (Role based only) (#5867)
hasura/graphql-engine: docs/graphql/core/api-reference/schema-metadata-api/remote-schema-permissions.rst

.. meta::
   :description: Manage remote schema permissions with the Hasura metadata API
   :keywords: hasura, docs, schema/metadata API, API reference, remote schema permissions, permission

.. _remote_schema_api_permission:

Schema/Metadata API Reference: Remote Schema Permissions
========================================================

.. contents:: Table of contents
   :backlinks: none
   :depth: 1
   :local:

Introduction
------------

The permission layer is designed to restrict the operations that can be
performed by various users. Permissions can be defined to only expose a certain

> part of the upstram remote schema for each role. By default, the ``admin`` role

Reviewer: ```suggestion
part of the upstream remote schema for each role. By default, the ``admin`` role
```

has unrestricted access to the upstream remote schema.

.. _add_remote_schema_permissions:

add_remote_schema_permissions
-----------------------------

> A remote schema permission is used to enforce constraints on the query that's

Reviewer 2: ```suggestion
A remote schema permission is used to enforce constraints on queries
```

> being queried to the remote schema. This is done by specifying the schema document

Reviewer 2: ```suggestion
to the remote schema. This is done by specifying the schema document
```

.. code-block:: graphql

  type User {
    user_id: Int
    name: String
    phone: String
    userMessages(whered: MessageWhereInpObj, includes: IncludeInpObj): [Message]
  }

  interface Communication {
    id: Int!
    msg: String!
  }

  type Message implements Communication {
    id: Int!
    name: String!
    msg: String!
    errorMsg: String
  }

  input MessageWhereInpObj {
    id: IntCompareObj
    name: StringCompareObj
  }

  input IntCompareObj {
    eq : Int
    gt : Int
    lt : Int
  }

  input StringCompareObj {
    eq : String
  }

  input IncludeInpObj {
    id: [Int]
    name: [String]
  }

  type Query {
    hello: String
    messages(where: MessageWhereInpObj, includes: IncludeInpObj): [Message]
    user(user_id: Int!): User
    users(user_ids: [Int]!): [User]
    message(id: Int!) : Message
  }

  type mutation_root {
    insert_user: (name: String!, phone: String!): User
  }

  schema {
    query: Query
    mutation: mutation_root
  }

Let's say we want to impose some restrictions on the ``user`` role:

1. Expose only the ``user_id``, ``name`` and the ``user_messages`` field in the ``User`` object.

> 2. Allow filtering of the messages only by ``name`` in the ``whered`` argument

Reviewer 2: ```suggestion
2. Allow filtering of the messages only by ``name`` in the ``where`` argument
```

   of the ``user_messages`` field.

> 3. Allow only ``hello``, ``messages`` and the ``user`` top level node in the ``Query`` object.

Reviewer 2: ```suggestion
3. Allow only the fields ``hello``, ``messages`` and the ``user`` top level node in the ``Query`` object.
```

           },
          "comment":"remote schema permissions for role: user"
       }
   }

.. _add_remote_schema_permissions_syntax:

Args syntax
^^^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - Key
     - Required
     - Schema
     - Description
   * - remote_schema
     - true
     - :ref:`RemoteSchemaName`
     - Name of the remote schema
   * - role
     - true
     - :ref:`RoleName`
     - Role
   * - definition
     - true
     - RemoteSchemaPermission_

>      - The permission definition

Reviewer 2: ```suggestion
     - The remote schema permission definition
```

   * - comment
     - false
     - text
     - Comment

.. _RemoteSchemaPermission:

RemoteSchemaPermission
&&&&&&&&&&&&&&&&&&&&&&

.. list-table::
   :header-rows: 1

   * - Key
     - Required
     - Schema
     - Description
   * - schema
     - true
     - GraphQL SDL
     - GraphQL SDL defining the role based schema

.. note::
   ``add_remote_schema_permissions`` will only work when the graphql-engine has enabled remote
   schema permissions. Remote schema permissions can be enabled by running the graphql-engine

>    with ``--enable-remote-schema-permissions`` server flag or by setting the   ``HASURA_GRAPHQL_ENABLE_REMOTE_SCHEMA_PERMISSIONS`` environment variable.

Reviewer 2: ```suggestion
   with the ``--enable-remote-schema-permissions`` server flag or by setting the   ``HASURA_GRAPHQL_ENABLE_REMOTE_SCHEMA_PERMISSIONS`` environment variable.
```
